File: project/backend-services/device-management-master-backend/infrastructure/endpoint/amqp/internal/src/main/java/pt/sensae/services/device/management/master/backend/infrastructure/endpoint/amqp/internal/mapper/DeviceInformationMapper.java
package pt.sensae.services.device.management.master.backend.infrastructure.endpoint.amqp.internal.mapper;

import org.springframework.stereotype.Service;
import pt.sensae.services.device.management.master.backend.application.DeviceDTO;
import pt.sensae.services.device.management.master.backend.application.DeviceEventMapper;
import pt.sensae.services.device.management.master.backend.application.DeviceNotificationDTO;
import pt.sensae.services.device.management.master.backend.application.ownership.DeviceWithAllOwnerDomains;
import pt.sensae.services.device.management.master.backend.application.ownership.DomainId;
import pt.sensae.services.device.management.master.backend.domain.model.DeviceInformation;
import pt.sensae.services.device.management.master.backend.domain.model.device.Device;
import pt.sensae.services.device.management.master.backend.domain.model.device.DeviceDownlink;
import pt.sensae.services.device.management.master.backend.domain.model.device.DeviceId;
import pt.sensae.services.device.management.master.backend.domain.model.device.DeviceName;
import pt.sensae.services.device.management.master.backend.domain.model.staticData.StaticDataLabel;
import pt.sensae.services.device.management.master.backend.infrastructure.endpoint.amqp.internal.model.*;

import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class DeviceInformationMapper implements DeviceEventMapper {

    @Override
    public DeviceNotificationDTO domainToUpdatedDto(DeviceInformation domain) {
        var info = new DeviceInformationDTOImpl();
        info.deviceId = domain.device().id().value().toString();
        info.name = domain.device().name().value();
        info.downlink = domain.device().downlink().value();
        if (info.downlink == null) info.downlink = "";

        info.records = domain.deviceRecords().entries().stream().map(e -> {
            var entry = new DeviceRecordEntryDTOImpl();
            entry.content = e.content();
            entry.label = e.label();
            return entry;
        }).collect(Collectors.toSet());

        info.staticData = domain.staticData().entries().stream().map(e -> {
            var entry = new DeviceStaticSensorEntryDTOImpl();
            entry.content = e.content();
            entry.label = modelToDto(e.label());
            return entry;
        }).collect(Collectors.toSet());

        info.subSensors = domain.subDevices().entries().stream().map(sub -> {
            var entry = new DeviceSubSensorDTOImpl();
            entry.subDeviceRef = sub.ref().value();
            entry.subDeviceId = sub.id().value().toString();
            return entry;
        }).collect(Collectors.toSet());

        info.commands = domain.commands().entries().stream().map(com -> {
            var entry = new DeviceCommandEntryDTOImpl();
            entry.id = com.id().value();
            entry.name = com.name().value();
            entry.port = com.port().value();
            entry.payload = com.payload().value();
            entry.subDeviceRef = com.ref().value();
            return entry;
        }).collect(Collectors.toSet());

        var notification = new DeviceNotificationDTOImpl();
        notification.deviceId = info.deviceId;
        notification.type = DeviceNotificationTypeDTOImpl.UPDATE;
        notification.information = info;
        return notification;
    }

    private DeviceStaticDataLabelDTOImpl modelToDto(StaticDataLabel model) {
        return switch (model) {
            case GPS_LATITUDE -> DeviceStaticDataLabelDTOImpl.GPS_LATITUDE;
            case GPS_LONGITUDE -> DeviceStaticDataLabelDTOImpl.GPS_LONGITUDE;
            case GPS_ALTITUDE -> DeviceStaticDataLabelDTOImpl.GPS_ALTITUDE;
            case BATTERY_MAX_VOLTS -> DeviceStaticDataLabelDTOImpl.BATTERY_MAX_VOLTS;
            case BATTERY_MIN_VOLTS -> DeviceStaticDataLabelDTOImpl.BATTERY_MIN_VOLTS;
            case MIN_DISTANCE -> DeviceStaticDataLabelDTOImpl.MIN_DISTANCE;
            case MAX_DISTANCE -> DeviceStaticDataLabelDTOImpl.MAX_DISTANCE;
        };
    }

    @Override
    public DeviceNotificationDTO domainToDeletedDto(DeviceId domain) {
        var notification = new DeviceNotificationDTOImpl();
        notification.deviceId = domain.value().toString();
        notification.type = DeviceNotificationTypeDTOImpl.DELETE;
        return notification;
    }

    @Override
    public Device dtoToDomain(DeviceDTO dto) {
        var device = (DeviceDTOImpl) dto;
        return new Device(new DeviceId(UUID.fromString(device.deviceId)), new DeviceName(device.name), new DeviceDownlink(device.downlink));
    }

    public DeviceWithAllOwnerDomains dtoToDomain(DeviceIdentityDTOImpl dto) {
        var domainIds = dto.information.owners.stream()
                .map(UUID::fromString)
                .map(DomainId::of)
                .collect(Collectors.toSet());
        return new DeviceWithAllOwnerDomains(DeviceId.of(UUID.fromString(dto.deviceId)), domainIds);
    }
}

File: project/frontend-services/libs/device-management/mapper/src/lib/mappers/DeviceInformationMapper.ts
import {
  DeviceInformationDTO,
  DeviceInformationResultDTO,
  SensorDataRecordLabelDTO,
} from '@frontend-services/device-management/dto';
import {
  Device,
  DeviceCommand,
  DeviceInformation,
  RecordEntry,
  SensorDataRecordLabel,
  StaticDataEntry,
  SubDevice
} from '@frontend-services/device-management/model';

export class DeviceInformationMapper {
  static dtoToModel(dto: DeviceInformationResultDTO): DeviceInformation {
    const records = dto.records.map((e) => new RecordEntry(e.label, e.content));

    const staticData = dto.staticData.map((e) => new StaticDataEntry(DeviceInformationMapper.labelDtoToModel(e.label), e.content));

    const subDevices = dto.subDevices.map(deviceRef => new SubDevice(deviceRef.ref, deviceRef.id));

    const commands = dto.commands.map(com => new DeviceCommand(com.id, com.name, com.payload, com.ref, com.port));

    const device = new Device(dto.device.id, dto.device.name, dto.device.downlink);

    const d = new Date(0); // The 0 there is the key, which sets the date to the epoch
    d.setUTCMilliseconds(dto.lastTimeSeen);
    return new DeviceInformation(device, records, staticData, subDevices, commands, d);
  }

  static modelToDto(model: DeviceInformation): DeviceInformationDTO {
    const records = model.records.map((e) => ({
      label: e.label,
      content: e.content,
    }));
    const staticData = model.staticData.map((e) => ({
      label: DeviceInformationMapper.labelModelToDto(e.label),
      content: e.content,
    }));

    const subDevices = model.subDevices.map(sub => ({
      id: sub.id,
      ref: sub.reference
    }));

    const commands = model.commands.map(com => ({
      id: com.id,
      name: com.name,
      payload: com.payload,
      ref: com.ref,
      port: com.port
    }));

    return {
      device: {id: model.device.id, name: model.device.name, downlink: model.device.downlink},
      records: records,
      subDevices: subDevices,
      staticData: staticData,
      commands: commands
    };
  }

  static labelModelToDto(model: SensorDataRecordLabel): SensorDataRecordLabelDTO {
    switch (model) {
      case SensorDataRecordLabel.BATTERY_MAX_VOLTS:
        return SensorDataRecordLabelDTO.BATTERY_MAX_VOLTS
      case SensorDataRecordLabel.BATTERY_MIN_VOLTS:
        return SensorDataRecordLabelDTO.BATTERY_MIN_VOLTS
      case SensorDataRecordLabel.MAX_DISTANCE:
        return SensorDataRecordLabelDTO.MAX_DISTANCE
      case SensorDataRecordLabel.MIN_DISTANCE:
        return SensorDataRecordLabelDTO.MIN_DISTANCE
      case SensorDataRecordLabel.GPS_ALTITUDE:
        return SensorDataRecordLabelDTO.GPS_ALTITUDE
      case SensorDataRecordLabel.GPS_LATITUDE:
        return SensorDataRecordLabelDTO.GPS_LATITUDE
      case SensorDataRecordLabel.GPS_LONGITUDE:
        return SensorDataRecordLabelDTO.GPS_LONGITUDE
      case SensorDataRecordLabel.ERROR:
        return SensorDataRecordLabelDTO.ERROR;
    }
  }

  static labelDtoToModel(model: SensorDataRecordLabelDTO): SensorDataRecordLabel {
    switch (model) {
      case SensorDataRecordLabelDTO.BATTERY_MAX_VOLTS:
        return SensorDataRecordLabel.BATTERY_MAX_VOLTS
      case SensorDataRecordLabelDTO.BATTERY_MIN_VOLTS:
        return SensorDataRecordLabel.BATTERY_MIN_VOLTS
      case SensorDataRecordLabelDTO.MAX_DISTANCE:
        return SensorDataRecordLabel.MAX_DISTANCE
      case SensorDataRecordLabelDTO.MIN_DISTANCE:
        return SensorDataRecordLabel.MIN_DISTANCE
      case SensorDataRecordLabelDTO.GPS_ALTITUDE:
        return SensorDataRecordLabel.GPS_ALTITUDE
      case SensorDataRecordLabelDTO.GPS_LATITUDE:
        return SensorDataRecordLabel.GPS_LATITUDE
      case SensorDataRecordLabelDTO.GPS_LONGITUDE:
        return SensorDataRecordLabel.GPS_LONGITUDE
      case SensorDataRecordLabelDTO.ERROR:
        return SensorDataRecordLabel.ERROR;
    }
  }
}
